integration/integration_test.go

```diff
@@ -0,0 +1,315 @@
+		t.Fatalf("Test executable failed to start.\nstderr:\n%s\nstdout:\n%s\n", stderr, stdout)
+	}
+}
+
+func TestServerFailsToStart(t *testing.T) {
+	var config strings.Builder
+
+	_, _ = config.WriteString("server:\n")
+	_, _ = config.WriteString("  strict_mode: false\n")
+	_, _ = config.WriteString("  port: 50052\n")
+	_, _ = config.WriteString("  tls: not_boolean\n")
+	_, _ = config.WriteString("logging:\n")
+	_, _ = config.WriteString("  level: debug\n")
+	_, _ = config.WriteString("  selectors: [\"*\"]\n")
+	_, _ = config.WriteString("  to_stderr: true\n")
+
+	env, err := NewTestingEnvironment(t, config.String())
+	if err != nil {
+		t.Fatalf("Error creating environment: %s", err)
+	}
+	t.Cleanup(func() { env.Stop() })
+
+	found, err := env.WaitUntil("stderr", "error unpacking shipper config")
+	if err != nil {
+		t.Fatalf("Error waiting for error message: %s", err)
+	}
+
+	if !found {
+		stderr, err := env.GetStderr()
+		if err != nil {
+			t.Errorf("Error getting standard error: %s", err)
+		}
+		stdout, err := env.GetStdout()
+		if err != nil {
+			t.Errorf("Error getting standard out: %s", err)
+		}
+		t.Fatalf("Test executable failed to start.\nstderr:\n%s\nstdout:\n%s\n", stderr, stdout)
+	}
+}
+
+func TestPublishMessage(t *testing.T) {
+	var config strings.Builder
+
+	_, _ = config.WriteString("server:\n")
+	_, _ = config.WriteString("  strict_mode: false\n")
+	_, _ = config.WriteString("  port: 50052\n")
+	_, _ = config.WriteString("  tls: false\n")
+	_, _ = config.WriteString("logging:\n")
+	_, _ = config.WriteString("  level: debug\n")
+	_, _ = config.WriteString("  selectors: [\"*\"]\n")
+	_, _ = config.WriteString("  to_stderr: true\n")
+
+	env, err := NewTestingEnvironment(t, config.String())
+	if err != nil {
+		t.Fatalf("Error creating environment: %s", err)
+	}
+	t.Cleanup(func() { env.Stop() })
+
+	found, err := env.WaitUntil("stderr", "gRPC server is ready and is listening on")
+	if err != nil {
+		t.Fatalf("Error waiting for server to start: %s", err)
+	}
+
+	if !found {
+		stderr, err := env.GetStderr()
+		if err != nil {
+			t.Errorf("Error getting standard error: %s", err)
+		}
+		stdout, err := env.GetStdout()
+		if err != nil {
+			t.Errorf("Error getting standard out: %s", err)
+		}
+		t.Fatalf("Test executable failed to start.\nstderr:\n%s\nstdout:\n%s\n", stderr, stdout)
+	}
+
+	client, err := env.NewClient("localhost:50052")
+	if err != nil {
+		t.Fatalf("Error creating client: %s", err)
+	}
+
+	unique := "UniqueStringToLookForInOutput"
+	sampleValues, err := helpers.NewStruct(map[string]interface{}{
+		"string": unique,
+		"number": 42,
+	})
+
+	e := &messages.Event{
+		Timestamp: timestamppb.Now(),
+		Source: &messages.Source{
+			InputId:  "input",
+			StreamId: "stream",
+		},
+		DataStream: &messages.DataStream{
+			Type:      "log",
+			Dataset:   "default",
+			Namespace: "default",
+		},
+		Metadata: sampleValues,
+		Fields:   sampleValues,
+	}
+	events := []*messages.Event{e}
+	ctx := context.Background()
```

Comment:
Should we use a context with a deadline propagated from `t` here?

In case the server gets stuck the test would fail with a timeout on the entire test, would be nice to know that this gRPC call timed out instead.

Reply:
That's a great idea.  added.